src/statistics/data.jl

```diff
@@ -48,21 +48,82 @@ quantile(d, p::T, s::Number) where {T<:Union{Number,AbstractVector}} = quantile(
 Spatial quantile of spatial data `sdata` at probability `p`.
 Optionally, specify the variable `v` and the block side `s`.
 """
-quantile(d::Data, p) = quantile(d, p, median_heuristic(d))
+quantile(d::Data, p) = quantile(d, p, mode_heuristic(d))
+
+function dist_matrix_random_sample(d, npoints=1000)
 
-function median_heuristic(d)
   # select at most 1000 points at random
   nel = nelements(d)
-  inds = sample(1:nel, min(nel, 1000), replace=false)
+  inds = sample(1:nel, min(nel, npoints), replace=false)
   X = (coordinates(centroid(d, ind)) for ind in inds)
-  D = pairwise(Euclidean(), X)
+  pairwise(Euclidean(), X)
+end
 
+bound_box_constr(d) = minimum(sides(boundingbox(domain(d))))
+
+function median_heuristic(d)
+  D = dist_matrix_random_sample(d)
   # median heuristic
   n = size(D, 1)
   m = median(D[i,j] for i in 1:n for j in 1:n if i > j)
 
-  # bounding box constraint
-  l = minimum(sides(boundingbox(domain(d))))
+  l = bound_box_constr(d)
+
+  min(m, l)
+end
+
+"""
+    mode_heuristic(d)
+
+Return the estimated mode of the inter-point distances for a set of locations.
+
+## References
+
+* Bickel & Frühwirth, 2005. [On a fast, robust estimator
+  of the mode: Comparisons to other robust estimators
+  with applications](https://doi.org/10.1016/j.csda.2005.07.011)
+"""
+function mode_heuristic(d)
+  D = dist_matrix_random_sample(d)
+  n = size(D, 1)
+  x_n = sort([D[i,j] for i in 1:n for j in 1:n if i > j])
+
+  while length(x_n) ≥ 4
+    n = length(x_n)
+    k = trunc(Int, ceil(n / 2) - 1)
+
+    inf = x_n[1:(n - k)]
+    sup = x_n[(k + 1):n]
+    diffs = sup - inf
```

Comment:
This difference is assuming that `sup` and `inf` will always have the same length. Are you sure this is always true? Worth double checking if this holds for `n` odd and even.